synapse/crypto/keyring.py

```diff
@@ -252,321 +251,225 @@ def verify_events_for_server(
-            """
-            if not verify_request.key_ids:
-                return defer.fail(
-                    SynapseError(
-                        400,
-                        "Not signed by %s" % (verify_request.server_name,),
-                        Codes.UNAUTHORIZED,
-                    )
-                )
 
-            logger.debug(
-                "Verifying %s for %s with key_ids %s, min_validity %i",
-                verify_request.request_name,
-                verify_request.server_name,
-                verify_request.key_ids,
-                verify_request.minimum_valid_until_ts,
+        if not verify_request.key_ids:
+            raise SynapseError(
+                400,
+                f"Not signed by {verify_request.server_name}",
+                Codes.UNAUTHORIZED,
             )
 
-            # add the key request to the queue, but don't start it off yet.
-            key_lookups.append(verify_request)
+        # Add they keys we need to verify to the queue for retrieval. We queue
+        # up requests for the same server so we don't end up with many in flight
+        # requests for the same keys.
+        found_keys_by_server = await self._server_queue.add_to_queue(
+            verify_request.to_fetch_key_request(), key=verify_request.server_name
+        )
 
-            # now run _handle_key_deferred, which will wait for the key request
-            # to complete and then do the verification.
-            #
-            # We want _handle_key_request to log to the right context, so we
-            # wrap it with preserve_fn (aka run_in_background)
-            return handle(verify_request)
+        # Since we batch up requests the returned set of keys may contain keys
+        # from other servers, so we pull out only the ones we care about.s
+        found_keys = found_keys_by_server.get(verify_request.server_name, {})
+
+        # For each signature to check we ensure we have fetched the necessary
+        # keys and the signature matches.
+        for key_id in verify_request.key_ids:
+            key_result = found_keys.get(key_id)
+            if not key_result:
+                raise SynapseError(
+                    401,
+                    f"Missing keys for {verify_request.server_name}: {key_id}",
+                    Codes.UNAUTHORIZED,
+                )
 
-        results = [process(r) for r in verify_requests]
+            if key_result.valid_until_ts < verify_request.minimum_valid_until_ts:
+                raise SynapseError(
+                    401,
+                    f"Failed to find key with recent enough `valid_until_ts` for {verify_request.server_name}: {key_id}",
+                    Codes.UNAUTHORIZED,
+                )
 
-        if key_lookups:
-            run_in_background(self._start_key_lookups, key_lookups)
+            verify_key = key_result.verify_key
+            try:
+                json_object = verify_request.get_json_object()
```

Comment:
has this moved inside the `try/except` deliberately? Not that I think it makes a lot of difference.

Reply:
No, and its a bit confusing that it is in there. I've moved it out